scripts/syntactics/common/utils.py

```diff
@@ -0,0 +1,424 @@
+                # The indices of the nodes that participate in the cycle
+                cycle = np.array(list(SCC))
+                # The probabilities of the current heads
+                old_heads = parse_preds[cycle]
+                old_head_probs = parse_probs[cycle, old_heads]
+                # Set the probability of depending on a non-head to zero
+                non_heads = np.array(list(dependents))
+                parse_probs[np.repeat(cycle, len(non_heads)), np.repeat([non_heads], len(cycle), axis=0).flatten()] = 0
+                # Get new potential heads and their probabilities
+                new_heads = np.argmax(parse_probs[cycle][:, tokens], axis=1) + 1
+                new_head_probs = parse_probs[cycle, new_heads] / old_head_probs
+                # Select the most probable change
+                change = np.argmax(new_head_probs)
+                changed_cycle = cycle[change]
+                old_head = old_heads[change]
+                new_head = new_heads[change]
+                # Make the change
+                parse_preds[changed_cycle] = new_head
+                tarjan.edges[new_head].add(changed_cycle)
+                tarjan.edges[old_head].remove(changed_cycle)
+        return parse_preds
+    else:
+        # block and pad heads
+        parse_probs = parse_probs * tokens_to_keep
+        parse_preds = np.argmax(parse_probs, axis=1)
+        return parse_preds
+
+
+def rel_argmax(rel_probs, length, ensure_tree=True):
+    """
+    adopted from Timothy Dozat https://github.com/tdozat/Parser/blob/master/lib/models/nn.py
+
+    :param rel_probs: seq_len x #rel
+    :param length: sequence length, a scalar
+    :param ensure_tree: constraint result to be a tree
+    :return: seq_len x 1 tensor
+    """
+    if ensure_tree:
+        rel_probs[:, ParserVocabulary.PAD] = 0
+        root = ParserVocabulary.ROOT
+        tokens = np.arange(1, length)
+        rel_preds = np.argmax(rel_probs, axis=1)
+        roots = np.where(rel_preds[tokens] == root)[0] + 1
+        if len(roots) < 1:
+            rel_preds[1 + np.argmax(rel_probs[tokens, root])] = root
+        elif len(roots) > 1:
+            root_probs = rel_probs[roots, root]
+            rel_probs[roots, root] = 0
+            new_rel_preds = np.argmax(rel_probs[roots], axis=1)
+            new_rel_probs = rel_probs[roots, new_rel_preds] / root_probs
+            new_root = roots[np.argmin(new_rel_probs)]
+            rel_preds[roots] = new_rel_preds
+            rel_preds[new_root] = root
+        return rel_preds
+    else:
+        rel_probs[:, ParserVocabulary.PAD] = 0
+        rel_preds = np.argmax(rel_probs, axis=1)
+        return rel_preds
+
+
+def reshape_fortran(tensor, shape):
+    """
+    The missing Fortran reshape for mx.NDArray
+    :param tensor: source tensor
+    :param shape: desired shape
+    :return: reordered result
+    """
+    return tensor.T.reshape(tuple(reversed(shape))).T
```

Comment:
We have adopted too many functions from https://github.com/tdozat/Parser .

Reply:
Probably true. Most of them are kept to ensure the consistency with original implementation. I only focused on the DL part. Shall I remove those unnecessary KMeans bucketing and Tarjan stuff? As the parser algorithm is good enough, the accuracy won't drop much.